{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "5b2dc925",
   "metadata": {},
   "source": [
    "### Overview\n",
    "This notebook imports the **marker genes** (a curated list of marker genes for each cell-type), extracts the data and convert them into CSV for the next annotation step (SCINA). \n",
    "\n",
    "*the marker genes were extracted from PanglaoDB [2] and Allen Brain Map [3], and placed in a CSV where each column contains the list of marker genes per cell-type. \n",
    "\n",
    "**This notebook is written in Python.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7a428c3c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-12-30T15:25:03.991049Z",
     "iopub.status.busy": "2022-12-30T15:25:03.990385Z",
     "iopub.status.idle": "2022-12-30T15:25:05.818287Z",
     "shell.execute_reply": "2022-12-30T15:25:05.817669Z"
    }
   },
   "outputs": [],
   "source": [
    "#Libraries and random state\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "%matplotlib inline "
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "3b8d6ea6",
   "metadata": {},
   "source": [
    "### Steps performed\n",
    "\n",
    "1. Import the marker genes CSV and the unannotated scRNA-seq\n",
    "\n",
    "2. Convert the gene names to Pascal case (or Upper Camel case). E.g. GENE1 to Gene1\n",
    "\n",
    "3. Get matching marker genes to the unannotated scRNA-seq\n",
    "\n",
    "4. Export filtered marker gene list"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "ec7c3b1a",
   "metadata": {},
   "source": [
    "#### Functions to handle case conversion and getting matched marker genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d56de300",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-12-30T15:25:05.822336Z",
     "iopub.status.busy": "2022-12-30T15:25:05.821955Z",
     "iopub.status.idle": "2022-12-30T15:25:05.887947Z",
     "shell.execute_reply": "2022-12-30T15:25:05.887348Z"
    }
   },
   "outputs": [],
   "source": [
    "#Function to convert gene names to Pascal case (or Upper Camel case). E.g. GENE1 to Gene1\n",
    "def convert_genes_to_Pascalcase(df_markers):\n",
    "    for (colName, colData) in df_markers.iteritems():\n",
    "        temp_colData = []\n",
    "\n",
    "        for s_gene in colData:\n",
    "            #Skip nan\n",
    "            if s_gene == \"NA\":\n",
    "                continue\n",
    "            #Lowercase all 2nd characters onwards\n",
    "            \n",
    "            if isinstance(s_gene, str):\n",
    "                s_gene = s_gene.title()\n",
    "                temp_colData.append(s_gene)\n",
    "        df_markers[colName]=temp_colData\n",
    "    return df_markers\n",
    "\n",
    "#Function to remove unmatched genes between the curated marker genes and the scRNA-seq to be annotated\n",
    "def get_matching_curated_markers(adata, df_markers):\n",
    "\n",
    "    df_markers = convert_genes_to_Pascalcase(df_markers)\n",
    "    print(df_markers)\n",
    "    df_markers_filtered = df_markers\n",
    "    d_removed_markers = df_markers\n",
    "\n",
    "    #Remove genes that are not found in the dataset\n",
    "    int_gene_count = sum(df_markers[df_markers != \"\"].count())\n",
    "    print(\"Before removing undetected genes: \" + str(int_gene_count) + \" genes\")\n",
    "\n",
    "    for (colName, colData) in df_markers.iteritems():\n",
    "        print(colName)\n",
    "        for s_gene in colData:\n",
    "            print(s_gene)\n",
    "            if adata.var_names.str.contains(s_gene).sum()==0:\n",
    "                df_markers_filtered = df_markers_filtered[df_markers_filtered[colName] != s_gene]\n",
    "            else:\n",
    "                d_removed_markers = d_removed_markers[d_removed_markers[colName] != s_gene]\n",
    "    \n",
    "    int_gene_count = sum(df_markers_filtered[df_markers_filtered != \"\"].count())\n",
    "    print(\"After removing undetected genes: \" + str(int_gene_count) + \" genes left\")\n",
    "    return df_markers_filtered, d_removed_markers"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "53b92cf2",
   "metadata": {},
   "source": [
    "#### Import the marker genes CSV and the unannotated scRNA-seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afd40eca",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import curated markergenes and unannotated scRNA-seq\n",
    "input_path_markergenes = \"../../data/demo_public/input/markergenes/PanglaoAllen_markers_gabagluta_nonneu.csv\"\n",
    "input_path_unannotatedscRNAseq =\"../../data/demo_public/input/AllenBrain_unannotated.h5ad\"\n",
    "\n",
    "#Read curated markergenes\n",
    "df_markergenes = pd.read_csv(input_path_markergenes,sep=\",\",index_col=False,na_filter=False, keep_default_na=False, na_values=['_'])\n",
    "print(\"Read curated markergenes\")\n",
    "print(df_markergenes)\n",
    "\n",
    "#Read unannotated scRNA-seq\n",
    "que_annotation = sc.read(input_path_unannotatedscRNAseq, cache=True)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "dbd02be9",
   "metadata": {},
   "source": [
    "#### Convert the gene names to Pascal case and get the matching marker genes to the unannotated scRNA-seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "53ef2fb3",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-12-30T15:25:05.892965Z",
     "iopub.status.busy": "2022-12-30T15:25:05.892644Z",
     "iopub.status.idle": "2022-12-30T15:25:08.580333Z",
     "shell.execute_reply": "2022-12-30T15:25:08.579770Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Read curated markergenes\n",
      "   Glutamatergic Gabaergic Non-Neuronal\n",
      "0        SLC17A7    SLC6A1           Qk\n",
      "1        SLC17A6    GABBR2       Zbtb20\n",
      "2          GRIN1      GAD1         Glul\n",
      "3         GRIN2B   GADD45B      Tsc22d4\n",
      "4            GLS      PAX2          Mt1\n",
      "..           ...       ...          ...\n",
      "95                                Gng12\n",
      "96                                 Aspa\n",
      "97                             Ppp1r14a\n",
      "98                              Fam107a\n",
      "99                                Phka1\n",
      "\n",
      "[100 rows x 3 columns]\n",
      "   Glutamatergic Gabaergic Non-Neuronal\n",
      "0        Slc17A7    Slc6A1           Qk\n",
      "1        Slc17A6    Gabbr2       Zbtb20\n",
      "2          Grin1      Gad1         Glul\n",
      "3         Grin2B   Gadd45B      Tsc22D4\n",
      "4            Gls      Pax2          Mt1\n",
      "..           ...       ...          ...\n",
      "95                                Gng12\n",
      "96                                 Aspa\n",
      "97                             Ppp1R14A\n",
      "98                              Fam107A\n",
      "99                                Phka1\n",
      "\n",
      "[100 rows x 3 columns]\n",
      "Before removing undetected genes: 125 genes\n",
      "Glutamatergic\n",
      "Slc17A7\n",
      "Slc17A6\n",
      "Grin1\n",
      "Grin2B\n",
      "Gls\n",
      "Meis2\n",
      "Slc17A8\n",
      "Slc1A1\n",
      "Slc1A2\n",
      "Slc1A6\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Gabaergic\n",
      "Slc6A1\n",
      "Gabbr2\n",
      "Gad1\n",
      "Gadd45B\n",
      "Pax2\n",
      "Slc32A1\n",
      "Gad2\n",
      "Vip\n",
      "Sst\n",
      "Pvalb\n",
      "Tnfaip8L3\n",
      "Sema3C\n",
      "Mybpc1\n",
      "Parm1\n",
      "Dlx1\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Non-Neuronal\n",
      "Qk\n",
      "Zbtb20\n",
      "Glul\n",
      "Tsc22D4\n",
      "Mt1\n",
      "Csrp1\n",
      "Hepacam\n",
      "Tmcc3\n",
      "S100A16\n",
      "Npas3\n",
      "Gatm\n",
      "Selenop\n",
      "Gng5\n",
      "Plekhg1\n",
      "Kcnj10\n",
      "Nfia\n",
      "St3Gal4\n",
      "Apoe\n",
      "Slc1A3\n",
      "Car2\n",
      "Maml2\n",
      "Olig1\n",
      "Plpp3\n",
      "Daam2\n",
      "Atp1A2\n",
      "Cdh20\n",
      "Clmn\n",
      "Clic4\n",
      "Dock10\n",
      "Cnp\n",
      "Pllp\n",
      "Dbi\n",
      "Erbin\n",
      "Gab1\n",
      "Gprc5B\n",
      "Scd2\n",
      "Cdh19\n",
      "Sox10\n",
      "Luzp2\n",
      "Gjc3\n",
      "Plcl1\n",
      "Frmd4B\n",
      "Rab31\n",
      "Sash1\n",
      "Ptprz1\n",
      "Tspan2\n",
      "Mal\n",
      "Ttyh2\n",
      "Plp1\n",
      "Ugt8A\n",
      "Bcas1\n",
      "Fa2H\n",
      "Ndrg1\n",
      "Cldn11\n",
      "Prr5L\n",
      "Mertk\n",
      "St18\n",
      "Cst3\n",
      "S100A1\n",
      "Plekhb1\n",
      "Erbb4\n",
      "Mag\n",
      "Tns3\n",
      "Gltp\n",
      "Slc6A1\n",
      "Mog\n",
      "St6Gal1\n",
      "Mt2\n",
      "Neat1\n",
      "S1Pr1\n",
      "Cd81\n",
      "Cryab\n",
      "Tmem88B\n",
      "Gpr37L1\n",
      "Epas1\n",
      "Phldb1\n",
      "Ermn\n",
      "Tcf7L2\n",
      "Bcan\n",
      "Megf10\n",
      "Gng11\n",
      "S100A13\n",
      "Zfp536\n",
      "Wscd1\n",
      "Arhgap31\n",
      "Creb5\n",
      "Gpc5\n",
      "Mobp\n",
      "Prex2\n",
      "Cobll1\n",
      "Timp3\n",
      "Sox6\n",
      "Carhsp1\n",
      "Hist1H2Bc\n",
      "Grb14\n",
      "Gng12\n",
      "Aspa\n",
      "Ppp1R14A\n",
      "Fam107A\n",
      "Phka1\n",
      "After removing undetected genes: 78 genes left\n"
     ]
    }
   ],
   "source": [
    "#Filter out unmatched genes \n",
    "df_markergenes_filtered, df_markergenes_removed = get_matching_curated_markers(que_annotation, df_markergenes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bfbd143d",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-12-30T15:25:08.585499Z",
     "iopub.status.busy": "2022-12-30T15:25:08.585162Z",
     "iopub.status.idle": "2022-12-30T15:25:08.590408Z",
     "shell.execute_reply": "2022-12-30T15:25:08.589809Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Removed markergenes\n",
      "  Glutamatergic Gabaergic Non-Neuronal\n",
      "3        Grin2B   Gadd45B      Tsc22D4\n"
     ]
    }
   ],
   "source": [
    "print(\"Removed markergenes\")\n",
    "print(df_markergenes_removed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "31844a6d",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-12-30T15:25:08.594221Z",
     "iopub.status.busy": "2022-12-30T15:25:08.593940Z",
     "iopub.status.idle": "2022-12-30T15:25:08.599558Z",
     "shell.execute_reply": "2022-12-30T15:25:08.598996Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Kept curated markergenes\n",
      "   Glutamatergic Gabaergic Non-Neuronal\n",
      "2          Grin1      Gad1         Glul\n",
      "4            Gls      Pax2          Mt1\n",
      "12                  Mybpc1         Gng5\n",
      "13                   Parm1      Plekhg1\n",
      "14                    Dlx1       Kcnj10\n",
      "..           ...       ...          ...\n",
      "92                              Carhsp1\n",
      "94                                Grb14\n",
      "95                                Gng12\n",
      "96                                 Aspa\n",
      "99                                Phka1\n",
      "\n",
      "[71 rows x 3 columns]\n"
     ]
    }
   ],
   "source": [
    "print(\"Kept curated markergenes\")\n",
    "print(df_markergenes_filtered)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3079523a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-12-30T15:25:08.602257Z",
     "iopub.status.busy": "2022-12-30T15:25:08.602001Z",
     "iopub.status.idle": "2022-12-30T15:25:08.611167Z",
     "shell.execute_reply": "2022-12-30T15:25:08.609930Z"
    }
   },
   "outputs": [],
   "source": [
    "#Export filtered marker genes\n",
    "output_path = \"../../data/demo_public/output/scina_filtered_markergenes.csv\"\n",
    "df_markergenes_filtered.to_csv(output_path, sep=\",\", index=False)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "4edafea0",
   "metadata": {},
   "source": [
    "#### References\n",
    "1. Chia, C. M., Roig Adam, A., & Moro, A. (2022). *In silico* multiple single-subject neural tissue screening using deconvolution on pseudo-bulk RNA-seq - a prototype. Bioinformatics and Systems Biology joint degree program. Vrije Universiteit Amsterdam and University of Amsterdam. \n",
    "\n",
    "2. Franzén, O., Gan, L. M., & Björkegren, J. L. M. (2019). PanglaoDB: a web server for exploration of mouse and human single-cell RNA sequencing data. Database : the journal of biological databases and curation, 2019, baz046. https://doi.org/10.1093/database/baz046\n",
    "\n",
    "3. Allen Institute for Brain Science (2004). Allen Mouse Brain Atlas, Mouse Whole Cortex and Hippocampus 10x. Available from mouse.brain-map.org. Allen Institute for Brain Science (2011)."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.13 ('rpy2env')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13 | packaged by conda-forge | (main, May 27 2022, 17:00:52) \n[Clang 13.0.1 ]"
  },
  "vscode": {
   "interpreter": {
    "hash": "488af5fec8fdf8c82286ebd3797745cedfd80ce7a1c80c975158a201a49ccec8"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
